loggingpy/log.py

```diff
@@ -0,0 +1,175 @@
+    '''
+
+    # Use this if you want to filter the path and line
+    # file_line = re.compile('.*File.*line.*')
+
+    @staticmethod
+    def exception_to_string(exception: Exception):
+        # traceback.extract_stack()[:-3] # prints the complete stack trace
+        stack = traceback.extract_tb(exception.__traceback__)  # add limit=??
+        pretty = traceback.format_list(stack)
+        return ''.join(pretty) + '\n  {} {}'.format(exception.__class__, exception)
+
+    @staticmethod
+    def hash_exception(exception: Exception):
+        stack_trace = LogEntryParser.exception_to_string(exception)
+        stack_trace_lines = stack_trace.split('\n')
+        # stack_trace_lines = [l for l in stack_trace.split('\n') if not LogEntryParser.file_line.match(l)]
+        stack_trace_lines = stack_trace_lines[:-2]
+        stack_trace_lines[-1] = stack_trace_lines[-1].split('(')[0]
+
+        # TBD: this routine would just create a single(32 character) hash of all exceptions:
+        # return hashlib.md5(''.join(stack_trace_lines).encode('utf-8'))
+
+        # if the list contains more than 10 hashes, trim it down by concatenating some
+        while len(stack_trace_lines) > 10:
+            stack_trace_lines[-2] = stack_trace_lines[-2] + stack_trace_lines[-1]
+            stack_trace_lines.pop(-1)
+
+        stack_hash = [hashlib.md5(s.encode('utf-8')).hexdigest() for s in stack_trace_lines]
+
+        # an MD5 hash has 32 characters - trim it down to 8 chars, risk of collisions is basically zero
+        # and it's wouldn't even be that much of a problem with proper logging (different context info, too)
+        hashes = [l[:8] for l in stack_hash]
+        return '.'.join(hashes)
+
+    @staticmethod
+    def parse_log_entry(log_entry: LogEntry):
+
+        exception = log_entry.exception
+        exception_info = None
+
+        if exception is not None:
+            exception_info = ExceptionInfo(exception_type=exception.__class__.__name__,
+                                           error_message=str(exception),
+                                           stack_trace=LogEntryParser.exception_to_string(exception),
+                                           exception_hash=LogEntryParser.hash_exception(exception))
+
+        # transparently wrap string values into an object to ensure a logged payload is always a JSON object rather
+        # than a scalar. We do not care about other primitives etc. If somebody is willing to log an int, it'll just
+        # get serialized
+        data = log_entry.data
+
+        if type(data) is str:
+            data = {"Message": data}
+
+        dto = {
+            "timestamp": log_entry.timestamp.isoformat(),
+            "level": log_entry.log_level.name,
+            "context": "" if log_entry.context is None else log_entry.context,
+            "payload_type": "" if log_entry.payload_type is None else log_entry.payload_type,
+            "data": data
```

Comment:
In .NET, I'm transforming the `data` key according to the payload type in order to avoid indexing collisions in ELK and other systems. (An element called `data` should always have the same structure, which isn't the case because we're logging a variety of different payloads).

Reply:
That's fine, I'm referring to the `data` property, that would lead to different data types called `data` and thus kill the indexer. Example of .NET transforming the key:

	{
		"payload_type": "CreditCardInfo",
		"credit_card_info": {
			"credit_card_number": "9183-6012-4978-8831",
			"credit_card_holder": "John Doe"
		}
	}